tests/notifier/notifier_test.py

```diff
@@ -44,3 +172,62 @@ def test_can_convert_created_at_datetime_to_timestamp_string(self):
 
         self.assertEquals(expected_timestamps,
                           converted_timestamps)
+
+    @mock.patch(
+        'google.cloud.forseti.notifier.notifier.find_notifiers', autospec=True)
+    @mock.patch(
+        'google.cloud.forseti.notifier.notifier.scanner_dao', autospec=True)
+    def test_no_notifications_for_empty_violations(
+        self, mock_dao, mock_find_notifiers):
+        """No notifiers are instantiated/run if there are no violations.
+
+        Setup:
+            Mock the scanner_dao and make its map_by_resource() function return
+            an empty violations map
+
+        Expected outcome:
+            The local find_notifiers() function is never called -> no notifiers
+            are looked up, istantiated or run."""
+        mock_dao.map_by_resource.return_value = dict()
+        mock_srvc_cfg = mock.MagicMock()
+        mock_srvc_cfg.get_global_config.return_value = FAKE_GLOBAL_CONFIGS
+        mock_srvc_cfg.get_notifier_config.return_value = FAKE_NOTIFIER_CONFIGS
+        notifier.run('iid-1-2-3', mock.MagicMock(), mock_srvc_cfg)
+        self.assertFalse(mock_find_notifiers.called)
+
+    @mock.patch(
+        ('google.cloud.forseti.notifier.notifiers.email_violations'
+         '.EmailViolations'), autospec=True)
+    @mock.patch(
+        'google.cloud.forseti.notifier.notifiers.gcs_violations.GcsViolations',
+        autospec=True)
+    @mock.patch(
+        'google.cloud.forseti.notifier.notifier.find_notifiers', autospec=True)
+    @mock.patch(
+        'google.cloud.forseti.notifier.notifier.scanner_dao', autospec=True)
+    def test_notifications_for_nonempty_violations(
+        self, mock_dao, mock_find_notifiers, mock_gcs_cls, mock_email_cls):
+        """The email/GCS upload notifiers are instantiated/run.
+
+        Setup:
+            Mock the scanner_dao and make its map_by_resource() function return
+            the FAKE_VIOLATIONS dict
+
+        Expected outcome:
+            The local find_notifiers() is called with with 'email_violations'
+            and 'gcs_violations' respectively. These 2 notifiers are
+            instantiated and run."""
+        mock_dao.map_by_resource.return_value = FAKE_VIOLATIONS
+        mock_srvc_cfg = mock.MagicMock()
+        mock_srvc_cfg.get_global_config.return_value = FAKE_GLOBAL_CONFIGS
+        mock_srvc_cfg.get_notifier_config.return_value = FAKE_NOTIFIER_CONFIGS
+
+        mock_email_obj = mock.MagicMock(spec=email_violations.EmailViolations)
+        mock_email_cls.return_value = mock_email_obj
```

Comment:
nit: please change this line to be as below, as it's more clear of what we are manipulating here.
```
mock_email_violations_cls.return_value = mock_email_violations
```

Reply:
Done.